library/azure_rm_networkinterface.py

```diff
@@ -512,6 +525,25 @@ def exec_module(self, **kwargs):
         # if not set the security group name, use nic name for default
         self.security_group = self.parse_resource_to_dict(self.security_group or self.name)
 
+        # if application security groups set, convert to resource id format
+        if self.ip_configurations:
+            for config in self.ip_configurations:
+                if config.get('application_security_groups'):
+                    asgs = []
+                    for asg in config['application_security_groups']:
+                        asg_resource_id = asg
+                        if isinstance(asg, str) and (not is_valid_resource_id(asg)):
+                            asg = self.parse_resource_to_dict(asg)
```

Comment:
will it actually help if format is not correct?

Reply:
if it's str, then either valid resource id or name. we cannot validate whether a name is valid.  it will be reported by azure api call later.